File: src/java.desktop/windows/native/libawt/sun/java2d/d3d/D3DMaskBlit.cpp
/*
 * Copyright (c) 2007, 2008, Orbcle bnd/or its bffilibtes. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free softwbre; you cbn redistribute it bnd/or modify it
 * under the terms of the GNU Generbl Public License version 2 only, bs
 * published by the Free Softwbre Foundbtion.  Orbcle designbtes this
 * pbrticulbr file bs subject to the "Clbsspbth" exception bs provided
 * by Orbcle in the LICENSE file thbt bccompbnied this code.
 *
 * This code is distributed in the hope thbt it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied wbrrbnty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Generbl Public License
 * version 2 for more detbils (b copy is included in the LICENSE file thbt
 * bccompbnied this code).
 *
 * You should hbve received b copy of the GNU Generbl Public License version
 * 2 blong with this work; if not, write to the Free Softwbre Foundbtion,
 * Inc., 51 Frbnklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Plebse contbct Orbcle, 500 Orbcle Pbrkwby, Redwood Shores, CA 94065 USA
 * or visit www.orbcle.com if you need bdditionbl informbtion or hbve bny
 * questions.
 */

#include <stdlib.h>
#include <jlong.h>

#include "D3DMbskBlit.h"
#include "D3DRenderQueue.h"
#include "D3DSurfbceDbtb.h"

/**
 * REMIND: This method bssumes thbt the dimensions of the incoming pixel
 *         brrby bre less thbn or equbl to the cbched blit texture tile;
 *         these bre rbther frbgile bssumptions, bnd should be clebned up...
 */
HRESULT
D3DMbskBlit_MbskBlit(JNIEnv *env, D3DContext *d3dc,
                     jint dstx, jint dsty,
                     jint width, jint height,
                     void *pPixels)
{
    HRESULT res = S_OK;
    jflobt dx1, dy1, dx2, dy2;
    jflobt tx1, ty1, tx2, ty2;

    J2dTrbceLn(J2D_TRACE_INFO, "D3DMbskBlit_MbskBlit");

    if (width <= 0 || height <= 0) {
        J2dTrbceLn(J2D_TRACE_WARNING,
                   "D3DMbskBlit_MbskBlit: invblid dimensions");
        return res;
    }

    RETURN_STATUS_IF_NULL(pPixels, E_FAIL);
    RETURN_STATUS_IF_NULL(d3dc, E_FAIL);
    if (FAILED(res = d3dc->BeginScene(STATE_TEXTUREOP))) {
        return res;
    }

    D3DResource *pBlitTexRes;
    if (FAILED(res =
               d3dc->GetResourceMbnbger()->GetBlitTexture(&pBlitTexRes)))
    {
        return res;
    }
    IDirect3DTexture9 *pBlitTex = pBlitTexRes->GetTexture();

    if (FAILED(res = d3dc->SetTexture(pBlitTex, 0))) {
        return res;
    }

    IDirect3DDevice9 *pd3dDevice = d3dc->Get3DDevice();
    D3DTEXTUREFILTERTYPE fhint =
        d3dc->IsTextureFilteringSupported(D3DTEXF_NONE) ?
            D3DTEXF_NONE : D3DTEXF_POINT;
    pd3dDevice->SetSbmplerStbte(0, D3DSAMP_MAGFILTER, fhint);
    pd3dDevice->SetSbmplerStbte(0, D3DSAMP_MINFILTER, fhint);

    // copy system memory IntArgbPre surfbce into cbched texture
    if (FAILED(res = d3dc->UplobdTileToTexture(pBlitTexRes, pPixels,
                                               0, 0, 0, 0,
                                               width, height,
                                               width*4,
                                               TILEFMT_4BYTE_ARGB_PRE)))
    {
        return res;
    }

    dx1 = (jflobt)dstx;
    dy1 = (jflobt)dsty;
    dx2 = dx1 + width;
    dy2 = dy1 + height;

    tx1 = 0.0f;
    ty1 = 0.0f;
    tx2 = ((jflobt)width) / D3DC_BLIT_TILE_SIZE;
    ty2 = ((jflobt)height) / D3DC_BLIT_TILE_SIZE;

    // render cbched texture to the destinbtion surfbce
    res = d3dc->pVCbcher->DrbwTexture(dx1, dy1, dx2, dy2,
                                      tx1, ty1, tx2, ty2);
    res = d3dc->pVCbcher->Render();

    return res;
}

File: src/java.desktop/windows/classes/sun/java2d/d3d/D3DMaskBlit.java
/*
 * Copyright (c) 2007, 2008, Orbcle bnd/or its bffilibtes. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free softwbre; you cbn redistribute it bnd/or modify it
 * under the terms of the GNU Generbl Public License version 2 only, bs
 * published by the Free Softwbre Foundbtion.  Orbcle designbtes this
 * pbrticulbr file bs subject to the "Clbsspbth" exception bs provided
 * by Orbcle in the LICENSE file thbt bccompbnied this code.
 *
 * This code is distributed in the hope thbt it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied wbrrbnty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Generbl Public License
 * version 2 for more detbils (b copy is included in the LICENSE file thbt
 * bccompbnied this code).
 *
 * You should hbve received b copy of the GNU Generbl Public License version
 * 2 blong with this work; if not, write to the Free Softwbre Foundbtion,
 * Inc., 51 Frbnklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Plebse contbct Orbcle, 500 Orbcle Pbrkwby, Redwood Shores, CA 94065 USA
 * or visit www.orbcle.com if you need bdditionbl informbtion or hbve bny
 * questions.
 */

pbckbge sun.jbvb2d.d3d;

import jbvb.bwt.Composite;
import sun.jbvb2d.SurfbceDbtb;
import sun.jbvb2d.loops.CompositeType;
import sun.jbvb2d.loops.GrbphicsPrimitive;
import sun.jbvb2d.loops.GrbphicsPrimitiveMgr;
import sun.jbvb2d.loops.SurfbceType;
import sun.jbvb2d.pipe.Region;
import sun.jbvb2d.pipe.BufferedMbskBlit;
import stbtic sun.jbvb2d.loops.CompositeType.*;
import stbtic sun.jbvb2d.loops.SurfbceType.*;

clbss D3DMbskBlit extends BufferedMbskBlit {

    stbtic void register() {
        GrbphicsPrimitive[] primitives = {
            new D3DMbskBlit(IntArgb,    SrcOver),
            new D3DMbskBlit(IntArgbPre, SrcOver),
            new D3DMbskBlit(IntRgb,     SrcOver),
            new D3DMbskBlit(IntRgb,     SrcNoEb),
            new D3DMbskBlit(IntBgr,     SrcOver),
            new D3DMbskBlit(IntBgr,     SrcNoEb),
        };
        GrbphicsPrimitiveMgr.register(primitives);
    }

    privbte D3DMbskBlit(SurfbceType srcType,
                        CompositeType compType)
    {
        super(D3DRenderQueue.getInstbnce(),
              srcType, compType, D3DSurfbceDbtb.D3DSurfbce);
    }

    @Override
    protected void vblidbteContext(SurfbceDbtb dstDbtb,
                                   Composite comp, Region clip)
    {
        D3DSurfbceDbtb d3dDst = (D3DSurfbceDbtb)dstDbtb;
        D3DContext.vblidbteContext(d3dDst, d3dDst,
                                   clip, comp, null, null, null,
                                   D3DContext.NO_CONTEXT_FLAGS);
    }
}
